exercises/pythagorean-triplet/example.py

```diff
@@ -1,37 +1,44 @@
-def triplets_in_range(min, max):
-    result = set([x for b in
-                  range(4, max + 1, 4) for x in
-                  primitive_triplets(b)])
-    for x, y, z in set(result):
-        a, b, c = (2 * x, 2 * y, 2 * z)
-        while c <= max:
-            if a >= min:
-                result.add((a, b, c))
-            a, b, c = (a + x, b + y, c + z)
-    return set([(d, e, f) for d, e, f in result if d >= min and f <= max])
-
-
-def maxMin(a, b):
-    return (a, b) if a > b else (b, a)
-
-
-def gcd(x, y):
-    while (y != 0):
-        x, y = y, x % y
-    return x
-
-
-def primitive_triplets(b):
-    if b % 4 != 0:
-        raise ValueError('b must be divisible by 4')
-    b2 = int(b / 2)
-    return set([(x[1], x[0], c) for x, c in
-                [(maxMin(m2 - n2, b), m2 + n2) for m, n, m2, n2 in
-                 [(m, n, int(m * m), int(n * n)) for m, n in
-                  set([maxMin(m, int(b2 / m)) for m in
-                       range(2, b2 + 1)
-                       if b2 % m == 0])
-                  if (m - n) % 2 == 1 and gcd(m, n) == 1]]])
+from math import sqrt, ceil
+try:
+    # Python 3
+    from math import gcd
+except ImportError:
+    # Python 2
+    from fractions import gcd
+
+
+def triplets_in_range(range_start, range_end):
+    for b in range(4, range_end + 1, 4):
+        for x, y, z in primitive_triplets(b):
+            a, b, c = (x, y, z)
+            while a < range_start:
+                a, b, c = (a + x, b + y, c + z)
+            while c <= range_end:
+                yield (a, b, c)
+                a, b, c = (a + x, b + y, c + z)
+
+
+def euclidian_coprimes(limit):
+    mn = limit // 2
+    for n in range(1, int(ceil(sqrt(mn)))):
+        if mn % n == 0:
+            m = mn // n
+            if (m - n) % 2 == 1 and gcd(m, n) == 1:
+                yield m, n
+
+
+def primitive_triplets(limit):
+    """See Euclid's formula
+    (https://en.wikipedia.org/wiki/Pythagorean_triple#Generating_a_triple)
+    for more information
+    """
+    for m, n in euclidian_coprimes(limit):
+        m2, n2 = m * m, n * n
+        a = m2 - n2
+        b = 2 * m * n
+        if a > b:
+            a, b = b, a
+        yield a, b, m2 + n2
```

Comment:
@yawpitch Picking up where we left off at https://github.com/exercism/python/pull/1599#discussion_r237600191

> keep the second sort it, it's more readable, and under the hood it's doing essentially the same thing.

> ` a, b, c = sorted((m2 - n2, 2 * m * n, m2 + n2))`
> `yield a, b, c`

I disagree. For one, including `c` in the sort is incorrect and redundant. Even excluding `c` from sorting, you end up with:

```Python
a, b = sorted((m2 - n2, 2 * m * n))
yield a, b, m2 + n2
```

...which, while less verbose that the current solution, is IMHO less readable.

Reply:
It's not really accurate to say that it's incorrect (you end up at a < b < c in all cases) and if I remember Timsort correctly there's no real redundancy (the first two items are swapped if necessary, then the third will be left untouched because it's already in the highest position) ... but I agree yours is clearer than the split sort.

That said, I much prefer to see the **c** actually labelled:

```python
a, b, c = m2 - n2, 2 * m * n, m2 + n2
if a > b:
    a, b = b, a
yield a, b, c
```
... perhaps?

Reply:
I can live with that. The slight overhead for an additional variable assignment per triplet is negligible, and it does make things much more readable.